### PR TITLE
HipblasLT and FusedMatMul

### PR DESCRIPTION
**Pull request by ekuznetsov139**

### CODE REVIEW

**Comment by jayfurmanek** [comment]
https://github.com/ROCmSoftwarePlatform/frameworks-internal/issues/2980

File: tensorflow/core/util/autotune_maps/conv_parameters.cc [hunk 1]
```diff
@@ -94,7 +94,7 @@ MatmulParameters::MatmulParameters(
   proto_.set_c_dtype(c_dtype);
 
   proto_.set_trans_a(trans_a);
-  proto_.set_trans_b(trans_a);
+  proto_.set_trans_b(trans_b);
```
Comment (wenchenvincent) [inline tensorflow/core/util/autotune_maps/conv_parameters.cc:97]: This seems to be fixing a pre-existing bug that's not related to hipBlasLt? If so, is it better to put it in a separate commit?

File: tensorflow/compiler/xla/stream_executor/rocm/rocm_driver.cc [hunk 2]
```diff
@@ -292,7 +292,10 @@ static tsl::Status InternalInit() {
   if (FLAGS_gpuexec_rocm_driver_inject_init_error) {
     LOG(ERROR) << "injecting ROCM init error; initialization will fail";
   } else {
+    static absl::Mutex mu_;
```
Comment (jayfurmanek) [inline tensorflow/compiler/xla/stream_executor/rocm/rocm_driver.cc]: Curious - did we not need this mutex for the hipInit() before?

File: tensorflow/compiler/xla/stream_executor/cuda/cuda_blas_lt.cc [hunk 3]
```diff
@@ -23,12 +23,18 @@ limitations under the License.
 #include <utility>
 #include <vector>
 
+#if GOOGLE_CUDA
 #include "third_party/gpus/cuda/include/cublasLt.h"
 #include "third_party/gpus/cuda/include/cublas_v2.h"
+#endif
 #include "tensorflow/compiler/xla/status_macros.h"
 #include "tensorflow/compiler/xla/stream_executor/blas.h"
+#if GOOGLE_CUDA
```
Comment (jayfurmanek) [inline tensorflow/compiler/xla/stream_executor/cuda/cuda_blas_lt.cc:32]: I suspect upstream won't like that we're adding these #if's throughout these cuda files, but having copies (rocm_blas_lt.cc) that are 90% the same is worse.

**Comment by jayfurmanek** [comment]
Does this also require setting this to true:

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/tensorflow/compiler/xla/debug_options_flags.cc#L76

**Comment by jayfurmanek** [comment]
retest Ubuntu-CPU please
retest Ubuntu-GPU-single please
retest Ubuntu-GPU-multi please

**Comment by jayfurmanek** [comment]
If you can rebase this one more time on top of the latest develop-upstream, everything should pass, then we can merge.
thx!

File: tensorflow/compiler/xla/service/gpu/conv_layout_normalization_test.cc [hunk 4]
```diff
@@ -109,10 +109,11 @@ ENTRY TestComputation {
 )";
 
   RunAndCompareWithLayoutsNormalized(hlo);
-
+#if !TENSORFLOW_USE_ROCM
   MatchOptimizedHlo(hlo, R"(
 // CHECK: (f32[8,32,4,5,5]{4,3,2,1,0}, u8[0]{0}) custom-call([[bitcast_8_0:%[^ ]+]], [[fusion_1:%[^ ]+]], [[bias_2:%[^ ]+]]), window={size=3x3x3 pad=1_1x1_1x1_1}, dim_labels=bf012_oi012->bf012, custom_call_target="__cudnn$convBiasActivationForward"
   )");
```
Comment (jayfurmanek) [inline tensorflow/compiler/xla/service/gpu/conv_layout_normalization_test.cc]: Is there an else case here for what appropriate on ROCM we can check?

File: tensorflow/compiler/xla/service/gpu/gpu_compiler.cc [hunk 5]
```diff
@@ -196,6 +196,11 @@ limitations under the License.
 
 namespace xla {
 namespace gpu {
+#if GOOGLE_CUDA
+using ComputeCap = se::CudaComputeCapability;
```
Comment (i-chaochen) [inline tensorflow/compiler/xla/service/gpu/gpu_compiler.cc:209]: maybe this is not needed anymore?

File: tensorflow/compiler/xla/service/gpu/conv_layout_normalization_test.cc (hunk 4 above)
Comment (ekuznetsov139) [reply tensorflow/compiler/xla/service/gpu/conv_layout_normalization_test.cc]: Not really. In case of CUDA, it verifies that the convolution is fused. On ROCM, I disable fusing for 3D convolutions (seems to be a MIOpen bug).

File: tensorflow/compiler/xla/service/gpu/gpu_compiler.cc (hunk 5 above)
Comment (ekuznetsov139) [reply tensorflow/compiler/xla/service/gpu/gpu_compiler.cc:209]: it allows the changes to lines 862-868

**Review by jayfurmanek — APPROVED** [review]

**Comment by sunway513** [comment]
can we move this PR to upstream then :-)

**Comment by jayfurmanek** [comment]
> can we move this PR to upstream then :-)

Yes, we'll push upstream once we have this merged in.

**Comment by jayfurmanek** [comment]
retest gpu-non-pip-multi please
retest gpu-pycpp please

**Comment by jayfurmanek** [comment]
retest cpu-non-pip please
retest gpu-pycpp please
